src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
+            size++;
+        } else {
+            checkIndex(index);
+            Node<T> myNode = tail;
+            for (int i = 0; i < index; i++) {
+                if (myNode.next != null) {
+                    myNode = myNode.next;
+                }
+            }
+            Node<T> prevNode = myNode.prev;
+            Node<T> nextNode = myNode;
+            Node<T> newNode = new Node<>(prevNode, value, nextNode);
+            nextNode.prev = newNode;
+            prevNode.next = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = tail;
+        for (int i = 0; i < index; i++) {
+            if (newNode != null && newNode.next != null) {
+                newNode = newNode.next;
+            }
+        }
+        return newNode.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nextNode = tail.next;
+        for (int i = 0; i < index; i++) {
+            if (nextNode.next != null) {
+                nextNode = nextNode.next;
+            }
+        }
+        Node<T> prevNode = nextNode.prev.prev;
+        Node<T> newNode = new Node<>(prevNode, value, nextNode);
+        T result = nextNode.prev.item;
+        if (index == 0) {
+            result = tail.item;
+            tail = newNode;
+        }
+        prevNode.next = newNode;
+        nextNode.prev = newNode;
+        return result;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> newNode = tail;
+        for (int i = 0; i < index; i++) {
+            if (newNode != null) {
+                newNode = newNode.next;
+            }
+        }
+        return unlink(newNode);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> newNode = tail;
+        for (int i = 0; i < size; i++) {
+            if (newNode.item == object || newNode.item != null && newNode.item.equals(object)) {
+                unlink(newNode);
+                return true;
+            } else if (newNode.next != null) {
+                newNode = newNode.next;
+            }
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index) {
+        if (index >= size || index < 0) {
+            throw new IndexOutOfBoundsException(
+                    "Wrong index, " + size);
+        }
+    }
+
+    private T unlink(Node<T> newNode) {
+        T result = newNode.item;
+        if (newNode.prev == head) {
+            tail = newNode.next;
```

Comment:
I can't understand this logic(

let's refactor code please
head and tail should be null by default (not empty nodes)
head is the first element, tail is the last element

head and tail should not be linked directly (nodes are NOT forming any circle)